File: Heap/Leetcode/Medium/TopKFrequentElements/TopKFrequentElements.cpp
// https://leetcode.com/problems/top-k-frequent-elements/
#include<bits/stdc++.h>
using namespace std;

vector<int> topKFrequent(vector<int> &nums, int k)
{
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
    unordered_map<int, int> map;

    for (int &ele : nums)
    {
        map[ele]++;
    }

    for (auto &pairs : map)
    {
        int freq = pairs.second;
        int ele = pairs.first;

        if (minHeap.size() < k)
        {
            minHeap.push({freq, ele});
        }
        else
        {
            if (minHeap.top().first < freq)
            {
                minHeap.pop();
                minHeap.push({freq, ele});
            }
        }
    }

    vector<int> ans;
    while (!minHeap.empty())
    {
        ans.emplace_back(minHeap.top().second);
        minHeap.pop();
    }

    return ans;
}

File: Heap/Leetcode/Medium/TopKFrequentElements/TopKFrequentElements.java
// https://leetcode.com/problems/top-k-frequent-elements/
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class TopKFrequentElements {
    class Pair {
        int first, second;
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> mp = new HashMap<>();

        PriorityQueue<Pair> minHeap = new PriorityQueue<>((a, b) -> {
            return a.first-b.first;
        });

        for(int ele: nums) {
            mp.compute(ele, (key, value) -> value == null ? 1 : value + 1); // map[ele] = freq
        }

        for(Map.Entry<Integer, Integer> e: mp.entrySet()) {
            int freq = e.getValue();
            int ele = e.getKey();

            if(minHeap.size() < k) {
                minHeap.offer(new Pair(freq, ele));
            } else {
                if(minHeap.peek().first < freq) {
                    minHeap.poll();
                    minHeap.offer(new Pair(freq, ele));
                }
            }
        }

        int[] arr = new int[k];
        int index = 0;
        while(!minHeap.isEmpty()) {
            arr[index++] = minHeap.poll().second;
        }
        return arr;
    }
}